test/test_deconvolutional_layer.h

```diff
@@ -41,164 +42,147 @@ TEST(deconvolutional, setup_nnp) {
+      0.1, -0.2,  0.3,
+      0.2, -0.3,  0.2
+  };
+
+  in = {
+      3.0, 2.0,
+      3.0, 0.0
+  };
 
-  weight[9]  = 0.0;  weight[10] = -0.1;  weight[11] = 0.1;
-  weight[12] = 0.1;  weight[13] = -0.2;  weight[14] = 0.3;
-  weight[15] = 0.2;  weight[16] = -0.3;  weight[17] = 0.2;
+  out_expected = {
+      0.9,  0.9,  0.8,  0.4,
+      0.9,  0.0,  0.1, -0.2,
+      0.15, -0.80, -0.55,  0.1,
+      0.15, -0.60,  0.15,  0.0,
 
-  in[0] = 3;  in[1] = 2;  in[2] = 3;  in[3] = 0;
+      0.0, -0.3,  0.1,  0.2,
+      0.3, -0.7,  0.8,  0.6,
+      0.9, -1.1,  0.9,  0.4,
+      0.6, -0.9,  0.6,  0.0
+  };
   // clang-format on
 
-  {
-    l.forward_propagation(in_data, out_data);
-
-    vec_t &out = out_tensor[0];
-    EXPECT_NEAR(0.900, out[0], 1E-5);
-    EXPECT_NEAR(0.900, out[1], 1E-5);
-    EXPECT_NEAR(0.800, out[2], 1E-5);
-    EXPECT_NEAR(0.400, out[3], 1E-5);
-    EXPECT_NEAR(0.900, out[4], 1E-5);
-    EXPECT_NEAR(0.000, out[5], 1E-5);
-    EXPECT_NEAR(0.100, out[6], 1E-5);
-    EXPECT_NEAR(-0.20, out[7], 1E-5);
-    EXPECT_NEAR(0.150, out[8], 1E-5);
-    EXPECT_NEAR(-0.80, out[9], 1E-5);
-    EXPECT_NEAR(-0.55, out[10], 1E-5);
-    EXPECT_NEAR(0.100, out[11], 1E-5);
-    EXPECT_NEAR(0.150, out[12], 1E-5);
-    EXPECT_NEAR(-0.60, out[13], 1E-5);
-    EXPECT_NEAR(0.150, out[14], 1E-5);
-    EXPECT_NEAR(0.000, out[15], 1E-5);
+  l.ith_parameter(0).set_data(Tensor<float_t>(weight));
+  l.forward({{in}}, out);
+  out_result = (*out[0])[0];
+
+  for (size_t i = 0; i < out_result.size(); i++) {
+    EXPECT_FLOAT_EQ(out_result[i], out_expected[i]);
   }
 }
 
-TEST(deconvolutional, fprop2) {
-  network<sequential> nn;
-
+TEST(deconvolutional, fprop_padding_same) {
   deconvolutional_layer l(2, 2, 3, 1, 2, padding::same);
+  l.setup(false);
 
   auto create_simple_tensor = [](size_t vector_size) {
     return tensor_t(1, vec_t(vector_size));
   };
 
-  tensor_t in_tensor     = create_simple_tensor(4),
-           out_tensor    = create_simple_tensor(32),
-           weight_tensor = create_simple_tensor(18),
-           bias_tensor   = create_simple_tensor(2);
+  tensor_t in_tensor  = create_simple_tensor(4),
+           out_tensor = create_simple_tensor(32);
 
   // short-hand references to the payload vectors
-  vec_t &in = in_tensor[0], &weight = weight_tensor[0];
+  vec_t weight = vec_t(18);
+  vec_t bias   = vec_t(2);
 
-  ASSERT_EQ(l.in_shape()[1].size(), 18u);  // weight
-
-  uniform_rand(in.begin(), in.end(), -1.0, 1.0);
+  uniform_rand(in_tensor[0].begin(), in_tensor[0].end(), -1.0, 1.0);
 
   std::vector<tensor_t *> in_data, out_data;
   in_data.push_back(&in_tensor);
-  in_data.push_back(&weight_tensor);
-  in_data.push_back(&bias_tensor);
   out_data.push_back(&out_tensor);
-  l.setup(false);
-  {
-    l.forward_propagation(in_data, out_data);
-    vec_t &out = out_tensor[0];
+  l.ith_parameter(0).set_data(Tensor<float_t>(weight));
+  l.ith_parameter(1).set_data(Tensor<float_t>(bias));
+  l.forward_propagation(in_data, out_data);
```

Comment:
for consistency can we use `forward(...)` and leave `forward_propagation(...)` for internals ?